:farmer: test_pub_sub, test_publisher and test_subscription failures

# Description

These tests have been failing for a long time in Iron connext and fastrtps-dynamic.

Reference builds:
* [Ici__nightly-fastrtps-dynamic_ubuntu_jammy_amd64#155](https://build.ros2.org/job/Ici__nightly-fastrtps-dynamic_ubuntu_jammy_amd64/155)
* [Ici__nightly-connext_ubuntu_jammy_amd64#158](https://build.ros2.org/job/Ici__nightly-connext_ubuntu_jammy_amd64/158/)


Test regressions:
* [test_tracetools.test_tracetools.test.test_subscription.TestSubscription.test_all](https://build.ros2.org/job/Ici__nightly-connext_ubuntu_jammy_amd64/158/testReport/junit/test_tracetools.test_tracetools.test.test_subscription/TestSubscription/test_all/)
* [test_tracetools.test_tracetools.test.test_publisher.TestPublisher.test_all](https://build.ros2.org/job/Ici__nightly-connext_ubuntu_jammy_amd64/158/testReport/junit/test_tracetools.test_tracetools.test.test_publisher/TestPublisher/test_all/)
* [test_tracetools.test_tracetools.test.test_pub_sub.TestPubSub.test_all](https://build.ros2.org/job/Ici__nightly-connext_ubuntu_jammy_amd64/158/testReport/junit/test_tracetools.test_tracetools.test.test_pub_sub/TestPubSub/test_all/)
* [projectroot.test_pub_sub](https://build.ros2.org/job/Ici__nightly-connext_ubuntu_jammy_amd64/158/testReport/junit/(root)/projectroot/test_pub_sub/)

All three contain the same log output:

Log output:
```
11: =================================== FAILURES ===================================
11: __________________________ TestSubscription.test_all ___________________________
11: /tmp/ws/install_isolated/tracetools_test/lib/python3.10/site-packages/tracetools_test/case.py:110: in setUp
11:     self.assertSetEqual(all_event_names, set(self._event_names))
11: E   AssertionError: Items in the first set but not the second:
11: E   'ros2:rmw_subscription_init'
11: E   'ros2:rmw_take'
11: ----------------------------- Captured stdout call -----------------------------
11: [INFO] [launch]: All log files can be found below /home/buildfarm/.ros/log/2024-08-19-01-58-16-902022-a9e85b8b6b09-444099
11: [INFO] [launch]: Default logging verbosity is set to INFO
11: [INFO] [tracetools_launch.action]: Writing tracing session to: /tmp/session-test-subscription-20240819015816
11: [INFO] [test_ping-1]: process started with pid [444104]
11: [INFO] [test_pong-2]: process started with pid [444106]
11: [test_pong-2] [INFO] [1724061498.067881993] [test_pong]: [output] some random ping string
11: [test_ping-1] [INFO] [1724061498.068240921] [test_ping]: [output] some random pong string
11: [test_pong-2] RTI Connext DDS Non-commercial license is for academic, research, evaluation and personal use only. USE FOR COMMERCIAL PURPOSES IS PROHIBITED. See RTI_LICENSE.TXT for terms. Download free tools at rti.com/ncl. License issued to Non-Commercial User license@rti.com For non-production use only.
11: [test_pong-2] Expires on 00-jan-00 See www.rti.com for more information.
11: [test_pong-2] spinning
11: [INFO] [test_pong-2]: process has finished cleanly [pid 444106]
11: [test_ping-1] RTI Connext DDS Non-commercial license is for academic, research, evaluation and personal use only. USE FOR COMMERCIAL PURPOSES IS PROHIBITED. See RTI_LICENSE.TXT for terms. Download free tools at rti.com/ncl. License issued to Non-Commercial User license@rti.com For non-production use only.
11: [test_ping-1] Expires on 00-jan-00 See www.rti.com for more information.
11: [test_ping-1] spinning
11: [INFO] [test_ping-1]: process has finished cleanly [pid 444104]
11: TRACE DIRECTORY: /tmp/session-test-subscription-20240819015816
11: - generated xml file: /tmp/ws/test_results/test_tracetools/test_subscription.xunit.xml -
11: 
11: ---------- coverage: platform linux, python 3.10.12-final-0 ----------
11: 
11: =========================== short test summary info ============================
11: FAILED test/test_subscription.py::TestSubscription::test_all - AssertionError...
11: ============================== 1 failed in 2.52s ===============================
11: -- run_test.py: return code 1
11: -- run_test.py: verify result file '/tmp/ws/test_results/test_tracetools/test_subscription.xunit.xml'
3/3 Test #11: test_subscription ................***Failed    3.22 sec



The following tests FAILED:
	  8 - test_pub_sub (Failed)
	  9 - test_publisher (Failed)
	 11 - test_subscription (Failed)
```

--- a/.github/workflows/sanity.yml
+++ b/.github/workflows/sanity.yml
@@ -3,7 +3,7 @@ on:
   pull_request:
   push:
     branches:
-      - rolling
+      - iron
   schedule:
     - cron: "0 5 * * *"
 jobs:
@@ -14,7 +14,7 @@ jobs:
       fail-fast: false
       matrix:
         distro:
-          - rolling
+          - iron
     steps:
     - uses: actions/checkout@v3
       with:
@@ -57,7 +57,7 @@ jobs:
       fail-fast: false
       matrix:
         distro:
-          - rolling
+          - iron
     steps:
     - uses: actions/checkout@v3
       with:

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -3,7 +3,7 @@ on:
   pull_request:
   push:
     branches:
-      - rolling
+      - iron
   schedule:
     - cron: "0 5 * * *"
 jobs:
@@ -16,25 +16,25 @@ jobs:
         include:
           # Normal build (binary)
           - os: ubuntu-22.04
-            distro: rolling
+            distro: iron
             build-type: binary
             instrumentation: instr-enabled
             tracepoints: tp-included
           # Normal build (source)
           - os: ubuntu-22.04
-            distro: rolling
+            distro: iron
             build-type: source
             instrumentation: instr-enabled
             tracepoints: tp-included
           # Build with instrumentation disabled
           - os: ubuntu-22.04
-            distro: rolling
+            distro: iron
             build-type: source
             instrumentation: instr-disabled
             tracepoints: tp-included
           # Normal build with tracepoints excluded
           - os: ubuntu-22.04
-            distro: rolling
+            distro: iron
             build-type: source
             instrumentation: instr-enabled
             tracepoints: tp-excluded

--- a/test_tracetools/CMakeLists.txt
+++ b/test_tracetools/CMakeLists.txt
@@ -133,6 +133,9 @@ if(BUILD_TESTING)
   # Only run tracing tests if instrumentation and tracepoints are included
   if(NOT TRACETOOLS_TRACEPOINTS_EXCLUDED)
     find_package(ament_cmake_pytest REQUIRED)
+    find_package(rmw_implementation_cmake REQUIRED)
+
+    # Tests to run with the default rmw implementation, which should not matter
     set(_test_tracetools_pytest_tests
       test/test_buffer.py
       test/test_executor.py
@@ -140,10 +143,7 @@ if(BUILD_TESTING)
       test/test_intra_pub_sub.py
       test/test_lifecycle_node.py
       test/test_node.py
-      test/test_pub_sub.py
-      test/test_publisher.py
       test/test_service.py
-      test/test_subscription.py
       test/test_timer.py
     )
     foreach(_test_path ${_test_tracetools_pytest_tests})
@@ -153,6 +153,34 @@ if(BUILD_TESTING)
         TIMEOUT 60
         WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
       )
+    endforeach()
+
+    # Tests to run with all instrumented/supported rmw implementations
+    set(_test_tracetools_pytest_tests_multi_rmw
+      test/test_pub_sub.py
+      test/test_publisher.py
+      test/test_subscription.py
+    )
+    set(_test_tracetools_rmw_implementations
+      rmw_cyclonedds_cpp
+      rmw_fastrtps_cpp
+    )
+    get_available_rmw_implementations(rmw_implementations)
+    foreach(_test_path ${_test_tracetools_pytest_tests_multi_rmw})
+      get_filename_component(_test_name ${_test_path} NAME_WE)
+      foreach(rmw_implementation ${_test_tracetools_rmw_implementations})
+        if(rmw_implementation IN_LIST rmw_implementations)
+          ament_add_pytest_test(${_test_name}__${rmw_implementation} ${_test_path}
+            ENV RMW_IMPLEMENTATION=${rmw_implementation}
+            APPEND_ENV PYTHONPATH=${CMAKE_CURRENT_BINARY_DIR}
+            TIMEOUT 60
+            WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+          )
+        else()
+          message(
+            "rmw implementation '${rmw_implementation}' not available for test '${_test_name}'")
+        endif()
+      endforeach()
     endforeach()
   endif()
 endif()

